crates/pallet-domains/src/block_tree.rs

```diff
@@ -0,0 +1,700 @@
+) -> Result<(), Error> {
+    let ExecutionReceipt {
+        consensus_block_number,
+        consensus_block_hash,
+        domain_block_number,
+        block_extrinsics_roots,
+        parent_domain_block_receipt_hash,
+        ..
+    } = execution_receipt;
+
+    if domain_block_number.is_zero() {
+        // The genesis receipt is generated and added to the block tree by the runtime upon domain
+        // instantiation, thus it is unchallengeable and must always be the same.
+        ensure!(
+            BlockTree::<T>::get(domain_id, domain_block_number).contains(&execution_receipt.hash()),
+            Error::BadGenesisReceipt
+        );
+    } else {
+        let execution_inbox = ExecutionInbox::<T>::get(domain_id, domain_block_number);
+        ensure!(
+            !block_extrinsics_roots.is_empty() && *block_extrinsics_roots == execution_inbox,
+            Error::InvalidExtrinsicsRoots
+        );
+    }
+
+    let excepted_consensus_block_hash =
+        frame_system::Pallet::<T>::block_hash(consensus_block_number);
+    ensure!(
+        *consensus_block_hash == excepted_consensus_block_hash,
+        Error::BuiltOnUnknownConsensusBlock
+    );
+
+    if let Some(parent_block_number) = domain_block_number.checked_sub(&One::one()) {
+        let parent_block_exist = BlockTree::<T>::get(domain_id, parent_block_number)
+            .contains(parent_domain_block_receipt_hash);
+        ensure!(parent_block_exist, Error::UnknownParentBlockReceipt);
+    }
+
+    match execution_receipt_type::<T>(domain_id, execution_receipt) {
+        ReceiptType::InFuture => {
+            log::error!(
+                "Unexpected in future receipt {execution_receipt:?}, which should result in \
+                `UnknownParentBlockReceipt` error as it parent receipt is missing"
+            );
+            Err(Error::InFutureReceipt)
+        }
+        ReceiptType::Pruned => {
+            log::error!(
+                "Unexpected pruned receipt {execution_receipt:?}, which should result in \
+                `InvalidExtrinsicsRoots` error as its `ExecutionInbox` is pruned at the same time"
+            );
+            Err(Error::PrunedReceipt)
+        }
+        ReceiptType::NewHead
+        | ReceiptType::NewBranch
+        | ReceiptType::CurrentHead
+        | ReceiptType::Stale => Ok(()),
+    }
+}
+
+/// Process the execution receipt to add it to the block tree
+///
+/// NOTE: only `NewHead`, `NewBranch` and `CurrentHead` type of receipt is expected
+/// for this function, passing other type of receipt will result in an `UnexpectedReceiptType`
+/// error.
+pub(crate) fn process_execution_receipt<T: Config>(
+    domain_id: DomainId,
+    submitter: OperatorId,
+    execution_receipt: ExecutionReceiptOf<T>,
+    receipt_type: ReceiptType,
+) -> Result<(), Error> {
+    let er_hash = execution_receipt.hash();
+    match receipt_type {
+        er_type @ ReceiptType::NewHead | er_type @ ReceiptType::NewBranch => {
```

Comment:
This does not look like rust 😉 
Did not know this existed. Might never use it 🤣